### PR TITLE
Stop supporting overwriting declared licenses

### PR DESCRIPTION
**Pull request by fviernau**
Removing or fixing broken declared licenses coming from the package meta data is supported by using a "declared license mapping" curation. Adding a declared license which wasn't declared in the original packages' meta data does not make sense,
because the curated package wouldn't reflect reality anymore. Therefore the overwriting is not needed as all valid declared license fixes can be achieved using the mapping approach. Since using a mapping is also superior in any regard remove the means to overwrite a declared license.

### CODE REVIEW

**Review by sschuberth — CHANGES_REQUESTED** [review]
I agree with the changes themselves, but I think we can squash them all into one commit and find a much shorter rationale to put into the commit message.

File: model/src/main/kotlin/PackageCurationData.kt [hunk 1]
```diff
@@ -46,7 +46,7 @@ data class PackageCurationData(
 
     /**
      * The concluded license as an [SpdxExpression]. It can be used to correct the license of a package in case the
-     * [declaredLicenses] found in the packages metadata or the licenses detected by a scanner do not match reality.
+     * declared licenses found in the packages metadata or the licenses detected by a scanner do not match reality.
```
[the remark this answers is not in the document]
Comment (fviernau) [reply model/src/main/kotlin/PackageCurationData.kt:43]: > The KDoc comment actually references the declared license of a package

I believe the comment intents to do that. Obviously it doesn't due to the brackets.

[the remark this answers is not in the document]
Comment (sschuberth) [reply model/src/main/kotlin/PackageCurationData.kt:43]: Ah, I was reading your commit message in reverse. How about rewording as follows:

```
PackageCurationsData: Remove an unintended reference to declaredLicenses in docs

The KDoc comment is supposed to refer to the declared licenses of a
package, not the `declaredLicenses` field of `PackageCurationsData` itself.
So reword the sentence to not use a reference link at all.```

[the remark this answers is not in the document]
Comment (sschuberth) [reply model/src/main/kotlin/PackageCurationData.kt:43]: Or better yet, just fix the link (and text):

```
    /**
     * The concluded license as an [SpdxExpression]. It can be used to correct the [declared licenses of a package]
     * [Package.declaredLicenses] in case the found in the packages metadata or the licenses detected by a scanner
     * do not match reality.
     */
```

**Comment by fviernau** [comment]
@sschuberth are internal changes needed on Bosch side before the merge? (e.g. in curations.yml)

...Thinking about adding commit on top which makes deserialization backwards compatible by means of an ignore, as soon as the tests are all green. Does that make sense?

**Comment by sschuberth** [comment]
> @sschuberth are internal changes needed on Bosch side before the merge? (e.g. in curations.yml)

No, luckily we didn't curate declared licenses so far, and personally I'm fine with not adding backwards compatibility here, as we should really force people to get rid of any curated declared licenses, as it's really dangerous to have them.

File: model/src/main/kotlin/PackageCurationData.kt [hunk 2]
```diff
@@ -45,8 +45,9 @@ data class PackageCurationData(
     val declaredLicenses: SortedSet<String>? = null,
 
     /**
-     * The concluded license as an [SpdxExpression]. It can be used to correct the license of a package in case the
-     * [declaredLicenses] found in the packages metadata or the licenses detected by a scanner do not match reality.
+     * The concluded license as an [SpdxExpression]. It can be used to correct the
+     * [declared licenses of a package][Package.declaredLicenses] in case the found in the packages metadata or the
```
Comment (sschuberth) [inline model/src/main/kotlin/PackageCurationData.kt:43]: Nit: You can have a line-break between `][` (exactly as in my suggested code for copy & paste) without breaking the link. That makes the formatting a bit nicer.

File: analyzer/src/main/kotlin/curation/ClearlyDefinedPackageCurationProvider.kt [hunk 3]
```diff
@@ -113,18 +111,11 @@ class ClearlyDefinedPackageCurationProvider(server: Server = Server.PRODUCTION)
...
-            // that contain "OTHER" as a license, also see https://github.com/clearlydefined/curated-data/issues/7836.
-            runCatching { declaredLicense.toSpdx(SpdxExpression.Strictness.ALLOW_CURRENT) }.getOrNull()?.toString()
-        }
-
         val sourceLocation = curation.described?.sourceLocation.toArtifactOrVcs()
 
         val pkgCuration = PackageCuration(
             id = pkgId,
             data = PackageCurationData(
-                declaredLicenses = declaredLicenseParsed?.let { sortedSetOf(it) },
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/ClearlyDefinedPackageCurationProvider.kt:127]: We cannot drop this without a substitute, but assign it to `concludedLicense` instead.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 4]
```diff
@@ -62,8 +62,6 @@ class Sw360PackageCurationProvider(sw360Configuration: Sw360StorageConfiguration
                     PackageCuration(
                         id = pkgId,
                         data = PackageCurationData(
-                            declaredLicenses = sw360Release.embedded?.licenses
-                                ?.mapNotNullTo(sortedSetOf()) { it?.shortName },
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:66]: We cannot drop this without a substitute, but assign it to `concludedLicense` instead.

File: model/src/main/kotlin/PackageCurationData.kt [hunk 5]
```diff
@@ -125,7 +119,7 @@ private fun applyCurationToPackage(targetPackage: CuratedPackage, curation: Pack
     } ?: base.vcs
 
     val authors = curation.authors ?: base.authors
-    val declaredLicenses = curation.declaredLicenses ?: base.declaredLicenses
+    val declaredLicenses = base.declaredLicenses
```
Comment (sschuberth) [inline model/src/main/kotlin/PackageCurationData.kt:122]: Can we please inline this variable?

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result-with-curations.yml [hunk 6]
```diff
@@ -459,12 +459,11 @@ analyzer:
         - "Steve Freeman"
         - "Tom Denley"
         declared_licenses:
-        - "curated license a"
```
[the remark this answers is not in the document]
Comment (fviernau) [reply analyzer/src/funTest/assets/projects/synthetic/gradle-all-dependencies-expected-result-with-curations.yml:462]: This commits message does not exist anymore after squashing it and shortening it (as you suggested in another comment).

**Comment by fviernau** [comment]
> I agree with the changes themselves, but I think we can squash them all into one commit and find a much shorter rationale to put into the commit message.

I did that and provided a shorter rationale. Please see if you like it now.

**Comment by sschuberth** [comment]
> > I agree with the changes themselves, but I think we can squash them all into one commit and find a much shorter rationale to put into the commit message.
> 
> I did that and provided a shorter rationale. Please see if you like it now.

Sorry, that comment of mine became obsolete by now, as it was based on my misunderstanding of the first commit. I liked the fact to have the KDoc fix separate. And I changed my mind about the second commit after another review: You were right that taking `declaredLicenseMapping` into account in the `DefaultLicenseInfoProvider` is anyway correct, independently of removing `declaredLicense` from `PackageCurationData`.

I'm really sorry for this misunderstanding, and I could understand if you'd want to keep this as-is now.

**Comment by fviernau** [comment]
> > > I agree with the changes themselves, but I think we can squash them all into one commit and find a much shorter rationale to put into the commit message.
> > 
> > 
> > I did that and provided a shorter rationale. Please see if you like it now.
> 
> Sorry, that comment of mine became obsolete by now, as it was based on my misunderstanding of the first commit. I liked the fact to have the KDoc fix separate. And I changed my mind about the second commit after another review: You were right that taking `declaredLicenseMapping` into account in the `DefaultLicenseInfoProvider` is anyway correct, independently of removing `declaredLicense` from `PackageCurationData`.
> 
> I'm really sorry for this misunderstanding, and I could understand if you'd want to keep this as-is now.

No worries, I've recovered these two commits to be separate.

File: analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt [hunk 7]
```diff
@@ -113,3 +117,10 @@ class Sw360PackageCurationProvider(sw360Configuration: Sw360StorageConfiguration
         return SW360ConnectionFactory().newConnection(sw360ClientConfig)
     }
 }
+
+private fun Collection<SW360SparseLicense>.toSpdx(): SpdxExpression? =
+    mapTo(mutableSetOf()) {
+        it.shortName.toSpdx()
+    }.reduceOrNull { left, right ->
+        SpdxCompoundExpression(left, SpdxOperator.AND, right)
+    }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/curation/Sw360PackageCurationProvider.kt:126]: How about this instead:

    DeclaredLicenseProcessor.process(mapTo(mutableSetOf()) { it.shortName}).spdxExpression

File: model/src/main/kotlin/licenses/DefaultLicenseInfoProvider.kt [hunk 8]
```diff
@@ -71,7 +71,7 @@ class DefaultLicenseInfoProvider(
                 licenses = pkg.declaredLicenses,
                 processed = pkg.declaredLicensesProcessed,
                 appliedCurations = curations.filter {
-                    it.curation.declaredLicenses != null || it.curation.declaredLicenseMapping.isNotEmpty()
+                    it.curation.declaredLicenseMapping.isNotEmpty()
                 }
```
Comment (sschuberth) [inline model/src/main/kotlin/licenses/DefaultLicenseInfoProvider.kt]: Nit: Do you want to make this a one-liner now again?

**Review by sschuberth — APPROVED** [review]

**Comment by sschuberth** [comment]
@fviernau, can be merged this soon? Because I'm planning to also make `comment` the first field in `PackageCurationData`, which I'd like to only start doing after this change.

**Review by sschuberth — DISMISSED** [review]

**Review by sschuberth — APPROVED** [review]
